### ISSUE TITLE
Drop Py3.6 support?

### ISSUE THREAD

**Issue by nbraud** [ppb/ppb-vector#184]
Now that Python 3.8 is out, and Python 3.7 is available starting in Debian 10 “Buster” (current stable) and Ubuntu 19.04 “Disco Dingo”, should we consider dropping Python 3.6 support?

Python 3.6 support currently relies on the `dataclasses` backport, which seems essentially-unmaintained since November 2018, and lacks [important bugfixes].  Those bugs are blocking for API improvements like #168.

[important bugfixes]: https://github.com/ericvsmith/dataclasses/pull/140

**Comment by astronouth7303** [comment]
This should track with PPB's own support cycle.

I think that's unofficially the last 2 major versions (so currently, 3.8 and 3.7), but I'm not sure that's been decided formally.

I'm going to just take maintainer's fiat and say that ppb-vector devel may drop 3.6. 1.0 exists and supports it, and that's good enough.

**Comment by nbraud** [comment]
@astronouth7303 Yeah, that was my recollection too (last 2 major releases of CPython), but I'd rather check with you and Piper first.

**Comment by nbraud** [comment]
Also, dropping Py3.6 from CI will conflict with all the CI updates that are in-flight, so that can wait a couple of days.

**Comment by astronouth7303** [comment]
Revisting some stuff. Do we want to drop 3.6 given pypy?

**Comment by nicoonoclaste** [comment]
> Revisting some stuff. Do we want to drop 3.6 given pypy?

PyPy has support for Py3.7 now, so I would be strongly in favour of dropping Py3.6 compatibility.
Though the `dataclasses` compat shim has been updated, so we could in principle keep support if we cared to.

**Comment by pathunstrom** [comment]
I'll note, 3.5 Python just hit EOL, 3.6 is still getting security patches, we should all have a conversation about support schedule sometime.

**Comment by nicoonoclaste** [comment]
> Though the `dataclasses` compat shim has been updated, so we could in principle keep support if we cared to.

Turns out, the maintainer of `dataclasses` (the PyPI package, not the stdlib module) merged the PR that contained the update, but failed to publish a new version...

**Comment by nicoonoclaste** [comment]
> I'll note, 3.5 Python just hit EOL, 3.6 is still getting security patches, we should all have a conversation about support schedule sometime.

I simply don't think that we can reasonably support people on Py3.6:
- Py3.6 support depends on the `dataclasses` compatibility shim; it is, as published on PyPI, mostly-broken, and its maintainer seems to be taking over 1.5 years to merge critical bugfixes... and still fail to publish new versions on PyPI.
- CPython 3.6 is not getting bugfixes since 2018.

Also, that discussion happened in the past, the outcome was “last 2 major versions”, and we referred to that above.

**Comment by AstraLuma** [comment]
Ok, 6mo later it might be time.

3.9, 3.8, and 3.7 are out and stable.

We can probably drop 3.6.

**Comment by AstraLuma** [comment]
Oh, dropping a major version of python is a breaking change, isn't it?

**Comment by nicoonoclaste** [comment]
> Oh, dropping a major version of python is a breaking change, isn't it?

Is it?  Py3.6 code should work just as well on Py3.7, and TBH I really do not think that the current situation is sensible, tying ourselves to Python releases that are 5 years old, haven't received bugfixes for the last 2+ years, and rely on a known-buggy, unmaintained compat shim.

**Comment by AstraLuma** [comment]
Oh, 3.6 is gonna be dropped upsteam in December iirc.

But it is a change that forces a class of users to make a change to their environment.
